### PR TITLE
Minimal support for arrays and matrices.

### LINKED ISSUE #535: Reify matrices
We need structs for matrices of compile-time size, templatized on the dimensions.
The naming convention should be `FixedMatrix`, `FixedStrictlyLowerTriangularMatrix`, etc.
They should be equipped with a 2-argument `operator()` for accessing entries; for accessing rows, choose between a 1-argument `operator()` and an `operator[]`.

### PR DESCRIPTION
**Pull request by pleroy**
Fix #535.

### CODE REVIEW

File: numerics/arrays.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+template<typename Scalar, int size>
+class FixedVector {
+ public:
+  FixedVector();  // Zero-initialized.
+  FixedVector(std::array<Scalar, size> const& data);
+  FixedVector(std::initializer_list<Scalar> const& data);
+
+  bool operator==(FixedVector const& right) const;
+
```
Comment (eggrobin) [inline numerics/arrays.hpp]: dd

File: numerics/arrays_body.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include "numerics/arrays.hpp"
+
+#include "glog/logging.h"
+
+namespace principia {
+namespace numerics {
+
+std::array<double, 2> const x{{1,2}};
```
Comment (eggrobin) [inline numerics/arrays_body.hpp]: wat

File: numerics/arrays_body.hpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+template<typename Scalar, int size>
+FixedVector<Scalar, size>::FixedVector() {
+  Scalar zero{};
+  data_.fill(zero);
+}
+
+template<typename Scalar, int size>
+FixedVector<Scalar, size>::FixedVector(std::array<Scalar, size> const& data)
+    : data_(data) {}
```
Comment (eggrobin) [inline numerics/arrays_body.hpp]: try removing that one.

File: numerics/arrays_body.hpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+bool FixedMatrix<Scalar, rows, columns>::operator==(
+    FixedMatrix const& right) const {
+  return data_ == right.data_;
+}
+
+template<typename Scalar, int rows, int columns>
+FixedMatrix<Scalar, rows, columns>&
+FixedMatrix<Scalar, rows, columns>::operator=(
+    std::initializer_list<Scalar> const& right) {
+  CHECK_EQ(size, right.size());
```
Comment (eggrobin) [inline numerics/arrays_body.hpp]: test it or remove it

File: numerics/arrays_body.hpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  std::copy(right.begin(), right.end(), data_.begin());
+  return *this;
+}
+
+template<typename ScalarLeft, typename ScalarRight, int rows, int columns>
+FixedVector<Product<ScalarLeft, ScalarRight>, rows> operator*(
+    FixedMatrix<ScalarLeft, rows, columns> const& left,
+    FixedVector<ScalarRight, columns> const& right) {
+  FixedVector<Product<ScalarLeft, ScalarRight>, rows> result;
+  auto const* row_start = left.data_.data();
```
Comment (eggrobin) [inline numerics/arrays_body.hpp]: just row. It's cleaner.

File: mathematica/newhall.nb [hunk 6]
```diff
@@ -10,10 +10,10 @@
...
 NotebookFileLineBreakTest
 NotebookDataPosition[       157,          7]
-NotebookDataLength[     25281,        640]
-NotebookOptionsPosition[     23665,        586]
-NotebookOutlinePosition[     24006,        601]
-CellTagsIndexPosition[     23963,        598]
+NotebookDataLength[     25825,        648]
+NotebookOptionsPosition[     24209,        594]
+NotebookOutlinePosition[     24550,        609]
+CellTagsIndexPosition[     24507,        606]
```
Comment (eggrobin) [inline mathematica/newhall.nb]: Update PDF.

**Comment by eggrobin** [comment]
This fixes #534.

File: numerics/arrays.hpp (hunk 1 above)
Comment (pleroy) [inline numerics/arrays.hpp]: Done.

File: numerics/arrays_body.hpp (hunk 2 above)
Comment (pleroy) [inline numerics/arrays_body.hpp]: Gone.

File: numerics/arrays_body.hpp (hunk 4 above)
Comment (pleroy) [inline numerics/arrays_body.hpp]: Done, and it caught this problem.

File: numerics/arrays_body.hpp (hunk 5 above)
Comment (pleroy) [inline numerics/arrays_body.hpp]: Done.

**Comment by eggrobin** [comment]
Jenkins, test this please.

**Comment by eggrobin** [comment]
Jenkins, test this please.

**Comment by eggrobin** [comment]
Jenkins, retest this please.

**Comment by eggrobin** [comment]
Well, I can't get Jenkins to rebuild the pull request, but a build using the SHA1 of the merge, namely 901c3f84078437b239267ff352c8c60c996cb04b, passes with VS2013 Update 4, so you can merge.
Please push your fixes to `numerics.vcxproj` (it can be in a separate CL, it has been broken since #526 anyway) so that things don't break too much with clang-cl.

**Comment by pleroy** [comment]
Merging now, will do the vcxproj later.

**Pull request merged**